enterprise/frontend/src/embedding-sdk/sdk-package/components/public/MetabaseProvider/MetabaseProvider.tsx

```diff
@@ -0,0 +1,114 @@
+import { getWindow } from "embedding-sdk/sdk-shared/lib/get-window";
+import type { SdkStore } from "embedding-sdk/store/types";
+import type { MetabaseProviderProps } from "embedding-sdk/types/metabase-provider";
+
+type MetabaseProviderInitDataWrapperProps = Pick<
+  MetabaseProviderProps,
+  "authConfig" | "allowConsoleLog"
+> & {
+  reduxStore: SdkStore;
+};
+
+/**
+ * We call `use-init-data` hook to initialize the SDK with the initial data.
+ * This is necessary when hooks are used before any SDK component is rendered.
+ */
+const MetabaseProviderInitDataWrapper = memo(function InitDataWrapper({
+  authConfig,
+  allowConsoleLog,
+  reduxStore,
+}: MetabaseProviderInitDataWrapperProps) {
+  const useInitData = window?.MetabaseEmbeddingSDK?.useInitData;
+
+  if (!reduxStore || !useInitData) {
+    throw new Error('Embedding SDK "useInitData" hook is not available');
+  }
+
+  useInitData({
+    reduxStore,
+    authConfig,
+    allowConsoleLog,
+  });
+
+  return null;
+});
+
+const MetabaseProviderInner = memo(function MetabaseProviderInner(
+  props: Omit<MetabaseProviderProps, "children">,
+) {
+  useLoadSdkBundle(props.authConfig.metabaseInstanceUrl);
+
+  const { isLoading } = useSdkLoadingState();
+
+  const {
+    props: { reduxStore: existingStore },
+  } = useMetabaseProviderPropsStore();
+
+  // Return existing store or create a new one
+  const reduxStore = useMemo(
+    () =>
+      isLoading
+        ? null
+        : (existingStore ?? getWindow()?.MetabaseEmbeddingSDK?.getSdkStore()),
+    [existingStore, isLoading],
+  );
+
+  const { initialized: metabaseProviderPropsStoreInitialized } =
+    useInitializeMetabaseProviderPropsStore(props, reduxStore);
+
+  useDeepCompareEffect(
+    function updateMetabaseProviderProps() {
+      ensureMetabaseProviderPropsStore().setProps(props);
+    },
+    [props],
+  );
+
+  if (!metabaseProviderPropsStoreInitialized || !reduxStore) {
+    return null;
+  }
+
+  return (
+    <MetabaseProviderInitDataWrapper
+      authConfig={props.authConfig}
+      allowConsoleLog={props.allowConsoleLog}
+      reduxStore={reduxStore}
+    />
+  );
+});
+
+/**
+ * A component that provides the Metabase SDK context and theme.
+ *
+ * @function
+ * @category MetabaseProvider
+ */
+export const MetabaseProvider = memo(function MetabaseProvider({
+  children,
+  ...props
+}: MetabaseProviderProps) {
+  return (
+    <ClientSideOnlyWrapper ssrFallback={children}>
+      <RenderSingleCopy
+        id="metabase-provider"
+        multipleRegisteredInstancesWarningMessage={
+          // eslint-disable-next-line no-literal-metabase-strings -- Warning message
+          "Multiple instances of MetabaseProvider detected. Metabase Embedding SDK may work unexpectedly. Ensure only one instance of MetabaseProvider is rendered at a time."
+        }
+      >
+        <MetabaseProviderInner {...props} />
+      </RenderSingleCopy>
```

Comment:
We must ensure that only a single inner provider is rendered, because it performs all initialization logic that must be done once, also the same for unmount cleanup logic